### DISCUSSION TITLE
How to apply Reflection filter?

### DISCUSSION THREAD

**Discussion by H1manshu21** [Blaze-Stars/Image-Filter-Application#26] [Q&A] [answered]
Discuss algorithm on how to apply `reflection` effect on `BMP` image.

**Comment by Ank1taS** [comment, marked as answer]
By applying the `reflection` effect a mirror image of the original image will be created. Pixels of the `BMP` image will be rearranged ie pixels on the right side of the image will end up on the left and vise-versa.

**Reply by H1manshu21** [reply]
> By applying the `reflection` effect a mirror image of the original image will be created. Pixels of the `BMP` image will be rearranged ie pixels on the right side of the image will end up on the left and vise-versa.

Attach examples to show original vs modified images.

**Reply by Ank1taS** [reply]
#### Example of Original Image vs Reflected (horizontally) Image

![vertical-mirror](https://user-images.githubusercontent.com/65461370/168765204-c831056c-05ae-4fce-be46-1c2dbcca8dc1.jpg)

#### Pseudocode for Reflection filter (Column wise)

```
function reflection (Argument1 rows, Argument2 columns, Argument3 pixel 2Darray ) {
        iterate and pick each row of the pixel 2D array
                iterate and pick each pixel of the row picked above
                        swap the chosen pixel with the corresponding pixel from another end(right) by 
                        taking a temporary pixel variable and then storing the values of 3 channels of the current pixel let x, 
                        store channels values of xth pixel from the left to the corresponding channels of the temporary pixel,
                        store channels values of xth pixel from the right to the corresponding channels of xth pixel from the left,
                        store channels values of temporary pixel to corresponding channels of xth pixel from the right.     
end
}
```

**Reply by H1manshu21** [reply]
@chandrakant100 @sabyasachi07 @Tushar7-coder @HembramBeta777 please have a look into this algorithm.

**Reply by Tushar7-coder** [reply]
Reflection or mirror of image can be calculated by swapping pixels of rows and columns .
For rows --> Swap pixels on horizontally(right to left or left to right).
for column--> Swap pixels on vertically (up to down ).

**Reply by H1manshu21** [reply]
> Reflection or mirror of image can be calculated by swapping pixels of rows and columns . For rows --> Swap pixels on horizontally(right to left or left to right). for column--> Swap pixels on vertically (up to down ).

Currently, we are planning to do it column-wise not row-wise. For row-wise we have to add an extra parameter and algorithm to implement it. But column-wise looks far better than row-wise.

**Reply by chandrakant100** [reply]
we will implement column wise first, later we can think of row wise.

**Reply by chandrakant100** [reply]
> Reflection or mirror of image can be calculated by swapping pixels of rows and columns . For rows --> Swap pixels on horizontally(right to left or left to right). for column--> Swap pixels on vertically (up to down ).

Good catch Thanks @Tushar7-coder for pointing this out.

**Comment by chandrakant100** [comment]
| Original Image and its Refelcted Version |
| ------------------------------------------- |
| <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/ef/Mount_Hood_reflected_in_Mirror_Lake%2C_Oregon.jpg/1280px-Mount_Hood_reflected_in_Mirror_Lake%2C_Oregon.jpg" height="300px" width="500px"> |

**Reply by H1manshu21** [reply]
This is a good example of reflection. 👍

**Reply by chandrakant100** [reply]
This examples is Top-Down reflection of Image as @Tushar7-coder said we have not planned to approach this feature we are aiming to reflect it left to right.

**Reply by chandrakant100** [reply]
The above examples given by @Ank1taS is now our current plan
